File: advance_health/advance_health/report/lead_final_report/lead_final_report.py
import frappe
from frappe.utils import getdate, nowdate
from datetime import datetime, timedelta


def execute(filters=None):
    filters = frappe._dict(filters or {})
    columns = get_columns()
    data = get_data(filters)
    return columns, data, None

def get_columns():
    columns = [
        {"label": "Source", "fieldname": "source", "fieldtype": "Data", "width": 200},
        {"label": "Total Lead Count", "fieldname": "total_lead_count", "fieldtype": "Int", "width": 150},
        {"label": "Converted Lead Count", "fieldname": "converted_lead_count", "fieldtype": "Int", "width": 150},
        {"label": "Interested Lead Count", "fieldname": "interested_lead_count", "fieldtype": "Int", "width": 150},
        {"label": "Overdue Lead Count", "fieldname": "overdue_lead_count", "fieldtype": "Int", "width": 150},
        {"label": "Lost Lead Count", "fieldname": "lost_lead_count", "fieldtype": "Int", "width": 150},
    ]
    return columns

def get_data(filters):
    # Get the duration and custom_created_at (date range) filters
    duration = filters.get('duration', 'All The Time')
    date_range = filters.get('custom_created_at', None)
    # print('date_rangeeeeeeeeeeeeeeeeeeeeeeee',date_range)
    where_conditions = []
    assign_to = filters.get('custom_assign_to', None)
    if assign_to:
        where_conditions.append(f"custom_assign_to IN ({', '.join(['%s' % frappe.db.escape(x) for x in assign_to])})")

    if duration != "All The Time":
        if duration == "Today":
            where_conditions.append(f"custom_created_on = '{nowdate()}'")
        elif duration == "Weekly":
            start_of_week = get_start_of_week()
            where_conditions.append(f"custom_created_on >= '{start_of_week}'")
        elif duration == "Monthly":
            first_day_of_month = get_first_day_of_month()
            where_conditions.append(f"custom_created_on >= '{first_day_of_month}'")
        elif duration == "Qtrly":
            quarter_start_date = get_quarter_start_date()
            where_conditions.append(f"custom_created_on >= '{quarter_start_date}'")
        elif duration == "Date Range Wise" and date_range:
            if isinstance(date_range, list) and len(date_range) == 2:
                start_date, end_date = date_range
                where_conditions.append(f"custom_created_on BETWEEN '{start_date}' AND '{end_date}'")
    
    # Add any additional filters here
    filters_condition = " AND ".join(where_conditions) if where_conditions else "1=1"

    # SQL query to fetch lead counts by source
    leads_data = frappe.db.sql(f"""
        SELECT source, status, COUNT(name) as count
        FROM `tabLead`
        WHERE {filters_condition}
        GROUP BY source, status
    """, as_dict=True)

    # Aggregate lead counts by source and status
    source_counts = {}
    for row in leads_data:
        source = row.source or "Other"
        status = row.status
        if source not in source_counts:
            source_counts[source] = {
                "total_lead_count": 0,
                "converted_lead_count": 0,
                "interested_lead_count": 0,
                "overdue_lead_count": 0,
                "lost_lead_count": 0,
            }
        source_counts[source]["total_lead_count"] += row.count
        if status == "Converted":
            source_counts[source]["converted_lead_count"] += row.count
        elif status == "Interested":
            source_counts[source]["interested_lead_count"] += row.count
        elif status == "Lost":
            source_counts[source]["lost_lead_count"] += row.count
        elif status == "Overdue":
            source_counts[source]["overdue_lead_count"] += row.count

    # Prepare the final data for the report
    data = []
    for source, counts in source_counts.items():
        row = {
            "source": source,
            **counts,
        }
        data.append(row)

    return data

# Helper functions for start of the week, month, and quarter
def get_start_of_week():
    today = datetime.today()
    start_of_week = today - timedelta(days=today.weekday())
    return start_of_week.date()

def get_first_day_of_month():
    today = datetime.today()
    return datetime(today.year, today.month, 1).date()

def get_quarter_start_date():
    today = datetime.today()
    month = today.month
    if month in [1, 2, 3]:
        return datetime(today.year, 1, 1).date()
    elif month in [4, 5, 6]:
        return datetime(today.year, 4, 1).date()
    elif month in [7, 8, 9]:
        return datetime(today.year, 7, 1).date()
    else:
        return datetime(today.year, 10, 1).date()

File: advance_health/advance_health/report/lead_final_report/lead_final_report.js
frappe.query_reports["Lead Final Report"] = {
    "filters": [
        {
            "fieldname": "duration",
            "label": __("Duration"),
            "fieldtype": "Select",
            "options": "\nAll The Time\nToday\nWeekly\nMonthly\nQtrly\nDate Range Wise",  // Added "Date Range Wise"
            "default": "All The Time",
            "on_change": function(query_report) {
                let duration = frappe.query_report.get_filter_value('duration');
                
                // Toggle visibility of date range filter based on the duration selection
                frappe.query_report.toggle_filter_display('custom_created_at', duration !== 'Date Range Wise');
                
                // Handle the filter logic based on the selected duration
                if (duration === 'All The Time') {
                    // Reset filter for custom_created_at
                    frappe.query_report.set_filter_value('custom_created_at', null);
                } else if (duration === 'Today') {
                    let today = frappe.datetime.get_today();
                    frappe.query_report.set_filter_value('custom_created_at', [today, today]);  // Set today's date
                } else if (duration === 'Weekly') {
                    let today = new Date();
                    let start_of_week = new Date(today.setDate(today.getDate() - today.getDay()));  // Start of the current week (Sunday)
                    let end_of_week = new Date(today.setDate(today.getDate() + 6));  // Saturday
                    frappe.query_report.set_filter_value('custom_created_at', [start_of_week, end_of_week]);
                } else if (duration === 'Monthly') {
                    let first_of_month = frappe.datetime.month_start(frappe.datetime.get_today());
                    let last_of_month = frappe.datetime.month_end(frappe.datetime.get_today());
                    frappe.query_report.set_filter_value('custom_created_at', [first_of_month, last_of_month]);
                } else if (duration === 'Qtrly') {
                    let quarter_start = get_quarter_start_date();
                    let quarter_end = frappe.datetime.add_days(quarter_start, 90);  // Approx 3 months
                    frappe.query_report.set_filter_value('custom_created_at', [quarter_start, quarter_end]);
                }
                
                // Refresh the report after setting the date range filter
                query_report.refresh();
            }
        },
        {
            "fieldname": "custom_created_at",
            "label": __("Created At"),
            "fieldtype": "DateRange",
            "hidden": 1,  // Initially hidden, will be shown when Date Range Wise is selected
            "on_change": function(query_report) {
                // Trigger the refresh when date range is changed
                query_report.refresh();
            }
        },
        {
			"fieldname": "custom_assign_to",
			"label": __("Assigned To"),
			"fieldtype": "MultiSelectList",
			"fieldtype": "MultiSelectList",
            "options": "User",
            "get_data": function (txt) {
                return frappe.db.get_link_options("User", txt, {
                    // Add any filters you need, for example:
                    // company: frappe.query_report.get_filter_value("company"),
                });
            },
		}
		
		
		
    ]
};

// Function to get the start date of the current quarter
function get_quarter_start_date() {
    let today = new Date();
    let month = today.getMonth() + 1; // Get current month (1-12)
    let quarter_start_month = Math.floor((month - 1) / 3) * 3 + 1; // Start of the current quarter
    return frappe.datetime.add_days(new Date(today.getFullYear(), quarter_start_month - 1, 1), 0);
}
